service/tablet_allocator.cc

```diff
@@ -12,12 +12,405 @@
 #include "replica/database.hh"
 #include "service/migration_manager.hh"
 #include "service/tablet_allocator.hh"
+#include "utils/stall_free.hh"
+#include "locator/load_sketch.hh"
+#include "utils/div_ceil.hh"
 
 using namespace locator;
 using namespace replica;
 
 namespace service {
 
+seastar::logger lblogger("load_balancer");
+
+class load_balancer {
+    using global_shard_id = tablet_replica;
+    using shard_id = seastar::shard_id;
+
+    struct shard_load {
+        size_t tablet_count;
+        std::unordered_set<global_tablet_id> candidates;
+
+        future<> clear_gently() {
+            return utils::clear_gently(candidates);
+        }
+    };
+
+    struct node_load {
+        host_id node;
+        uint64_t shard_count = 0;
+        uint64_t tablet_count = 0;
+
+        // The maximum per-shard load in terms of tablet count after the node would be internally balanced.
+        uint64_t balanced_max_load = 0;
+
+        std::vector<shard_id> candidates; // heap which tracks most-loaded shards using candidates_cmp().
+        std::vector<shard_load> shards;
+
+        // Call when tablet_count changes.
+        void update() {
+            balanced_max_load = div_ceil(tablet_count, shard_count);
+        }
+
+        auto candidates_cmp() {
+            return [this] (const auto& a, const auto& b) {
+                return shards[a].tablet_count < shards[b].tablet_count;
+            };
+        }
+
+        future<> clear_gently() {
+            return utils::clear_gently(shards);
+        }
+    };
+
+    token_metadata_ptr _tm;
+public:
+    load_balancer(token_metadata_ptr tm)
+        : _tm(std::move(tm)) {
+    }
+
+    future<migration_plan> make_plan() {
+        const locator::topology& topo = _tm->get_topology();
+        migration_plan plan;
+
+        // Prepare plans for each DC separately and combine them to be executed in parallel.
+        for (auto&& dc : topo.get_datacenters()) {
+            auto dc_plan = co_await make_plan(dc);
+            lblogger.info("Prepared {} migrations in DC {}", dc_plan.size(), dc);
+            std::copy(dc_plan.begin(), dc_plan.end(), std::back_inserter(plan));
+        }
+
+        lblogger.info("Prepared {} migrations", plan.size());
+        co_return std::move(plan);
+    }
+
+    future<migration_plan> make_plan(sstring dc) {
+        /// The algorithm aims to equalize tablet count on each shard.
+        /// This goal is based on the assumption that every shard has similar processing power and space capacity,
+        /// and that each tablet has equal consumption of those resources. So by equalizing tablet count per shard we
+        /// equalize resource utilization.
+        ///
+        /// The algorithm produces a migration plan which is a set of instructions about which tablets to move
+        /// where. The plan is a small increment, not a complete plan. To achieve balance, the algorithm should
+        /// be invoked iteratively until an empty plan is returned.
```

Comment:
Does the plan have to be fully executed before we start another one? Or can we keep generating plans until we run out of things to do (given current transitions)?

Reply:
Ack